### PR TITLE
[WIP] perf improvements?

### PR DESCRIPTION
**Pull request by GavinJoyce**
possible perf improvements? we'll see...

### CODE REVIEW

File: lib/mongo-parser-rb/field.rb [hunk 1]
```diff
@@ -7,7 +7,6 @@ def initialize(field)
     end
 
     def value_in_document(document)
-      document = stringify_keys(document)
```
Comment (GavinJoyce) [inline lib/mongo-parser-rb/field.rb]: no need to `stringify_keys` the document multiple times

File: lib/mongo-parser-rb/field.rb [hunk 2]
```diff
@@ -22,8 +21,6 @@ def value_in_document(document)
     end
 
     def in_document?(document)
-      document = stringify_keys(document)
```
Comment (GavinJoyce) [inline lib/mongo-parser-rb/field.rb]: no need to `stringify_keys` the document multiple times

**Comment by ciaranlee** [comment]
interesting!

**Comment by ciaranlee** [comment]
these look pretty safe to :ship: I reckon?

**Comment by GavinJoyce** [comment]
I'd like to split this up into much smaller changes, the `stringify_keys` change will be small but will have the biggest impact.

I'll keep spiking and cut a small PR on Monday

**Comment by GavinJoyce** [comment]
Tests are now passing and I've added a perf test with `or` predicates:

**before**:
```
➜  mongo-parser-rb git:(gj/perf) ✗ ruby performance/perf.rb
       user     system      total        real
simple query parsing 14.460000   0.060000  14.520000 ( 14.628961)
simple query matching 59.600000   0.230000  59.830000 ( 60.255688)
```

**after**:
```
➜  mongo-parser-rb git:(gj/perf-improvements) ✗ ruby performance/perf.rb
       user     system      total        real
simple query parsing 12.500000   0.060000  12.560000 ( 12.658346)
simple query matching 26.690000   0.130000  26.820000 ( 27.011667)
```

**Comment by ciaranlee** [comment]
Nice! Sounds like a good plan

File: lib/mongo-parser-rb/query/expression.rb [hunk 3]
```diff
@@ -2,71 +2,17 @@ module MongoParserRB
...
-          equality_operator?(operator) || conjunction_operator?(operator) || inversion_operator?(operator) || elemMatch_operator?(operator)
-        end
-
-      end
+      INVERSION_OPERATORS = [:$not].freeze
+      CONJUNCTION_OPERATORS = [:$and, :$or].freeze
+      NEGATIVE_EQUALITY_OPERATORS = [:$nin, :$ne].freeze
+      EQUALITY_OPERATORS = [:$eq, :$gt, :$lt, :$gte, :$lte, :$in, :$nin, :$ne].freeze
+      ELEM_MATCH_OPERATORS = [:$elemMatch].freeze
+      ALL_OPERATORS = [:$eq, :$gt, :$lt, :$gte, :$lte, :$in, :$nin, :$ne, :$and, :$or, :$not, :$elemMatch].freeze
```
Comment (dehora) [inline lib/mongo-parser-rb/query/expression.rb]: > I'd like to split this up into much smaller changes,

This set looks good to go in another pr

Comment (GavinJoyce) [inline lib/mongo-parser-rb/query/expression.rb]: Yep, I'll cut two PRs in the morning

**Comment by GavinJoyce** [comment]
I've split these changes into three PRs:

 * https://github.com/intercom/mongo-parser-rb/pull/10
 * https://github.com/intercom/mongo-parser-rb/pull/12
 * https://github.com/intercom/mongo-parser-rb/pull/13
